### ISSUE TITLE
Difficulty compiling Linux kernel for ni/nilrt/nilrt/19.1 and ni/linux/nilrt/19.1/4.14

### ISSUE THREAD

**Issue by chrisb2244** [ni/nilrt#15]
I'm having some problems compiling the NILRT kernel for LabVIEW 2019 SP1 (which I'm hoping is the 19.1 branches in this and the ni/linux repositories) in order to enable an additional kernel module.

Initially I tried using a Linux machine I had easy access to without using the nilrt build-chain, this used a GCC-8.3.1 compiler and after using `make oldconfig` and `make menuconfig` having copied the running configuration from a cRIO-9045 (using `zcat /proc/config.gz > existingConfig` and then various scp commands to get a .config file in the linux directory on the build machine).

When running `make modules_prepare` in this configuration, I get an error from SE_Linux:
```
In file included from scripts/selinux/genheaders/genheaders.c:19:
./security/selinux/include/classmap.h:245:2: error: #error New address family defined, please update secclass_map.
 #error New address family defined, please update secclass_map.
```
Some searching seems to tell me that this is because my build machine's glibc version is newer than the intended version (or the previously built version, or the version for this config?)

I then attempted to use the nilrt build-chain, but ran into some issues with missing commands. I'll probably come back to this, but without root access I can imagine rpcgen and chrpath might be fiddly to get correctly picked up in scripts (maybe not...)

I then downloaded the toolchain mentioned in the README from https://www.ni.com/en-gb/support/downloads/software-products/download.gnu-c---c---compile-tools-x64.html#338443
These installed without any problems but I haven't had any success using them; the closest attempt gives the following error:
```
arch/x86/Makefile:245: *** You are building kernel with non-retpoline compiler, please update your compiler..  Stop.
```
which I'm guessing is because although `gcc --version` still returns 8.3.1 (system version), `${CROSS_COMPILE}gcc --version` now gives `x86_64-nilrt-linux-gcc (GCC) 6.3.0`, and versions older than 8, or 7.x with backported patches, apparently don't have whichever patch satisfies the kernel in relation to some branch target injection mitigation.

Presumably there is some correct way to go about this - which path is best for me to move down to best accomplish my goal (in this case, compiling the NAT tables parts of netfilter either into the kernel or as a module)?

**Comment by chaitu236** [comment]
Yes, nilrt/19.1/4.14 branch in ni/linux is correct.

The kernel sources already contain a defconfig (nati_x86_64_defconfig) that can be used instead of needing to copy config file from target.

I haven't tested gcc 8.3.1 specifically but was able to compile the kernel by running **make nati_x86_64_defconfig** and **make** with gcc 7.4.0 and gcc 6.3.0, Some of our team members also had success compiling with gcc version newer than 8.3.1.

If that doesn't work on your machine, there is a Dockerfile in https://github.com/ni/nilrt/tree/nilrt/master/sumo/scripts/docker (that installs gcc 6.3.0 at this time) that I verified works.

**Comment by chrisb2244** [comment]
Thank you for the information about the Dockerfile - building that and installing a few extra packages (libncurses-devel for menuconfig, libelf-dev, squashfs-tools) allowed me to successfully build the kernel and modules.

Loading them by following the instructions in KERNEL_SOURCE also worked smoothly and after rebooting into safe mode to disable the startup application (which I think caused some lockup due to the NI drivers not having been reversioned when I first rebooted) I managed to get everything booting, running as before and with a new module compiled.

Thanks for the help!
